### PR TITLE
merge dev into master

### PR DESCRIPTION
**Pull request by 5im1z**
changes include functionalized FP, added relevant globals in config, changes primary_site into collection_site, new function in terra.py that's called in FP function, and several one-off task notebooks

### CODE REVIEW

**Review by jkobject — COMMENTED** [review]
for ccle_tasks/update-wes-workspace.ipynb, can you add in the title, wha type of update it was?

File: depmapomics/config_dev.py [hunk 1]
```diff
@@ -304,6 +304,22 @@
 WESCNWORKSPACE = "broad-firecloud-ccle/DepMap_WES_CN_hg38"
 WESMUTWORKSPACE = "broad-firecloud-ccle/DepMap_Mutation_Calling_CGA_pipeline"
 
+
+############## Fingerprinting
+
+# Local directory to save intermediate files to
+WORKING_DIR = "temp/"
+
+FPWORKSPACE = "broad-firecloud-ccle/CCLE_SNP_QC"
```
Comment (jkobject) [inline depmapomics/config_dev.py:313]: wouldn't we want to use some temp workpaces for the dev branch? (same for WESCNWORKSPACE/ WESMUTWORKSPACE )

File: depmapomics/copynumbers.py [hunk 2]
```diff
@@ -408,10 +410,10 @@ def _CCLEPostProcessing(
         )
         wessegments, genecn, wesfailed = postProcess(
             wesrefworkspace,
-            AllSamplesetName if AllSamplesetName else samplesetname,
+            setEntity="pair_set",
```
Comment (jkobject) [inline depmapomics/copynumbers.py:413]: let's make this a parameter of the _CCLEPostProcessing function instead (and a config.py constant)

File: depmapomics/copynumbers.py [hunk 3]
```diff
@@ -467,10 +469,10 @@ def _CCLEPostProcessing(
     )
     wgssegments, genecn, wgsfailed = postProcess(
         wgsrefworkspace,
-        AllSamplesetName if AllSamplesetName else samplesetname,
+        setEntity="sample_set",
```
Comment (jkobject) [inline depmapomics/copynumbers.py:472]: same here.

File: depmapomics/copynumbers.py [hunk 4]
```diff
@@ -467,10 +469,10 @@ def _CCLEPostProcessing(
...
+        sampleset=AllSamplesetName if AllSamplesetName else samplesetname,
         todrop=todropwgs,
         save_output=folder,
-        segmentsthresh=2000,
```
Comment (jkobject) [inline depmapomics/copynumbers.py:473]: I think in the end we decided to keep the threshold for WGS, didn't we?

File: depmapomics/fingerprinting.py [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+from genepy import terra
+from genepy.utils import helper as h
+from depmapomics import tracker
+from depmapomics.config import *
+from depmapomics import terra as myterra
+
+
+def updateLOD(
+    wm, sampleset, taiga_dataset, taiga_filename, working_dir, upload_to_taiga=True
+):
```
Comment (jkobject) [inline depmapomics/fingerprinting.py:18]: let's add documentation now. :)

File: depmapomics/fingerprinting.py [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+            index="LEFT_SAMPLE", columns="RIGHT_SAMPLE", values="LOD_SCORE"
+        )
+        new_lod_list.append(lod_mat)
+    new_lod_mat = pd.concat(new_lod_list)
+    new_lod_mat.index.name = None
+    new_lod_mat = new_lod_mat.T
+
+    # Update LOD matrix ( have to update (A+a)*(B+b) = (AB)+(aB)+(Ab)+(ab))
+    tc = TaigaClient()
+    prev_lod_mat = tc.get(name=taiga_dataset, file=taiga_filename)
```
Comment (jkobject) [inline depmapomics/fingerprinting.py:41]: let's move this out of the function to keep it usable by everyone

File: depmapomics/fingerprinting.py [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+from depmapomics import terra as myterra
+
+
+def updateLOD(
+    wm, sampleset, taiga_dataset, taiga_filename, working_dir, upload_to_taiga=True
+):
+    # Here we update the fingerprint LOD matrix on taiga with the new fingerprints
+    # Generate matrix with LOD score for new fingerprint vcfs
+    new_lod_list = []
+    sample_batch_pair_df = wm.get_entities("sample_batch_pair")
```
Comment (jkobject) [inline depmapomics/fingerprinting.py:22]: let's maake the entity a parameter too

File: depmapomics/fingerprinting.py [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+        wm.upload_entities("sample_batch_pair", pair_df)
+    except:
+        print("still can't update sample_batch_pair")
+        # in case it does not work
+        pair_df.to_csv(working_dir + "sample_batch_pairs.tsv", sep="\t")
+        if h.askif(
+            "Please upload the file ../sample_batch_pairs.tsv to the terra workspace as a new data table, and type yes once \
+      finished, else write no to quit and they will not be added"
+        ):
+            print("sample_batch_pair manually updated")
```
Comment (jkobject) [inline depmapomics/fingerprinting.py:235]: else:
   raise the error

File: depmapomics/fingerprinting.py [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+    # update sample_batch_pair_set to include all sample_batch_pairs
+    try:
+        wm.upload_entities("sample_batch_pair_set", sample_batch_pair_set_df)
+    except:
+        print("still can't update sample_batch_pair_set")
+        # in case it does not work
+        sample_batch_pair_set_df.to_csv(
+            working_dir + "sample_batch_pair_set.tsv", sep="\t"
+        )
+        if h.askif(
```
Comment (jkobject) [inline depmapomics/fingerprinting.py:257]: same herre

File: depmapomics/fingerprinting.py [hunk 10]
```diff
@@ -0,0 +1,357 @@
+# fingerprinting.py
+
+import pandas as pd
+import numpy as np
+import dalmatian as dm
+from taigapy import TaigaClient
+from genepy.google import gcp
+import asyncio
```
Comment (jkobject) [inline depmapomics/fingerprinting.py:8]: do we use it?

File: depmapomics/release.py [hunk 11]
```diff
@@ -457,7 +455,6 @@ def make(
             """
 
 ## RNAseq Omics:
-
 NEW LINES:
```
Comment (jkobject) [inline depmapomics/release.py:458]:
```suggestion

NEW LINES:
```

File: depmapomics/release.py [hunk 12]
```diff
@@ -230,9 +230,7 @@ def make(
             "# "
             + val
             + """ dataset:
```
Comment (jkobject) [inline depmapomics/release.py:232]:
```suggestion
            + """ dataset:

```

File: depmapomics/release.py [hunk 13]
```diff
@@ -230,9 +230,7 @@ def make(
...
-
 ## DNAseq Omics:
-
 NEW LINES:
```
Comment (jkobject) [inline depmapomics/release.py:234]:
```suggestion

NEW LINES:
```

File: depmapomics/loading.py [hunk 14]
```diff
@@ -1174,7 +1174,7 @@ def update(
         add_to_samplesets (list, optional): add new samples to additional sample_sets on terra. Defaults to []
     """
     # uploading to our bucket (now a new function)
-    samples = terra.changeToBucket(
+    terra.changeToBucket(
```
Comment (jkobject) [inline depmapomics/loading.py:1177]:
```suggestion
    samples =  terra.changeToBucket(
```
we need this right?

File: depmapomics/fingerprinting.py [hunk 15]
```diff
@@ -0,0 +1,357 @@
...
+        (
+            updated_lod_mat.loc[new_ids.union(old_ids), old_ids],
+            new_lod_mat.transpose().loc[new_ids.union(old_ids, new_ids)],
+        ),
+        axis=1,
+    )
+    updated_lod_mat.to_csv(working_dir + taiga_filename + ".csv")
+
+    # Upload updated LOD matrix to Taiga
+    if upload_to_taiga:
```
Comment (jkobject) [inline depmapomics/fingerprinting.py:57]: same here. let's keep this in the notebook for now so the function is more generic

File: depmapomics/fingerprinting.py [hunk 16]
```diff
@@ -0,0 +1,357 @@
...
+                    "encoding": "utf-8",
+                }
+            ],
+            add_all_existing_files=True,
+        )
+    return new_ids, updated_lod_mat
+
+
+def checkMismatches(v, ref, samples, thr=100):
+    should = {}
```
Comment (jkobject) [inline depmapomics/fingerprinting.py:75]: I know I use crazy bad variable in my notebooks when I do some WIP stuff but for a functionalized code, let's make this better here:
maybe at least: [v, res, should,] should be updated (in the next function too)

File: depmapomics/fingerprinting.py [hunk 17]
```diff
@@ -0,0 +1,357 @@
...
+        sampleset ([type]): [description]
+        sid ([type]): [description]
+        working_dir ([type]): [description]
+        bamcolname ([type]): [description]
+        workspace ([type], optional): [description]. Defaults to WORKSPACE.
+
+    Author:
+        William Colgan (wcolgan@broadinstitute.org)
+    """
+
```
Comment (jkobject) [inline depmapomics/fingerprinting.py:293]: this is where we import taiga.

File: depmapomics/fingerprinting.py [hunk 18]
```diff
@@ -0,0 +1,357 @@
...
+    sid = "id"
```
Comment (jkobject) [inline depmapomics/fingerprinting.py:294]: should be a parameter of the function

File: depmapomics/fingerprinting.py [hunk 19]
```diff
@@ -0,0 +1,357 @@
...
+    samples = pd.concat([rnasamples, wgssamples])
+    bams = samples[bamcolname]
+    bams[sid] = bams.index
+    print("adding " + str(len(bams)) + " new samples to the fingerprint")
+    wm = dm.WorkspaceManager(workspace).disable_hound()
+
+    # Upload sample sheet
+    samples_df = pd.DataFrame()
```
Comment (jkobject) [inline depmapomics/fingerprinting.py:302]:
```suggestion
```

File: depmapomics/fingerprinting.py [hunk 20]
```diff
@@ -0,0 +1,357 @@
...
+            working_dir + "sample_batch_pair_set.tsv", sep="\t"
+        )
+        if h.askif(
+            "Please upload the file ../sample_batch_pair_set.tsv to the terra workspace as a new data table, and type yes once \
+      finished, else write no to quit and they will not be added"
+        ):
+            print("sample_batch_pair_set manually updated")
+
+
+async def _CCLEFingerPrint(
```
Comment (jkobject) [inline depmapomics/fingerprinting.py:264]: we should do like in the other tasks and have a generic function and a CCLE specific one.. that calls the generic + do the CCLE only things (like using taiga)

File: depmapomics/terra.py [hunk 21]
```diff
@@ -226,5 +229,109 @@ def copyToWorkspace(
...
+                wm.namespace,
+                wm.workspace,
+                update["entityType"],
+                update["name"],
+                update["operations"],
+            )
+            if r.status_code == 200:
+                print("Successfully updated {}.".format(update["name"]))
+            elif r.status_code >= 400:
+                print("Unable to update entity attributes")
```
Comment (jkobject) [inline depmapomics/terra.py:334]:
```suggestion
                raise BlockingIOError("Unable to update entity attributes")
```

File: depmapomics/terra.py [hunk 22]
```diff
@@ -226,5 +229,109 @@ def copyToWorkspace(
...
+                    )
+                )
+            else:
+                print(
+                    "Successfully updated attribute '{}' for {} {}s.".format(
+                        attrs.name, len(attrs), etype
+                    )
+                )
+        elif r.status_code >= 400:
+            raise print("Unable to update entity attributes")
```
Comment (jkobject) [inline depmapomics/terra.py:317]:
```suggestion
            raise BlockingIOError("Unable to update entity attributes")
```

File: depmapomics/terra.py [hunk 23]
```diff
@@ -226,5 +229,109 @@ def copyToWorkspace(
...
+        if addMissing and len(unmatched) > 0:
+            print("found these columns to be missing in workspace: " + str(unmatched))
+            track = tracker[tracker.index.isin(unmatched)][columns].rename(
+                columns=rename
+            )
+            track.index.name = "sample_id"
+            wm.upload_samples(track)
+
+
+def updateReferences(wm, etype, attrs):
```
Comment (jkobject) [inline depmapomics/terra.py:253]: wow, nice! 
Just need to add some docstring :)

File: depmapomics/terra.py [hunk 24]
```diff
@@ -226,5 +229,109 @@ def copyToWorkspace(
         print("found these columns to be missing in workspace: " + str(unmatched))
         track = tracker[tracker.index.isin(unmatched)][columns].rename(columns=rename)
         track.index.name = "sample_id"
-        wm.upload_samples(track)
+        miss = set(columns) - set(sam.columns)
+        if len(track) == 0 and not addMissing:
```
Comment (jkobject) [inline depmapomics/terra.py:233]: addMissing is always true

File: depmapomics/terra.py [hunk 25]
```diff
@@ -226,5 +229,109 @@ def copyToWorkspace(
...
+            raise ValueError("wrong tracker or index non matching")
+        unmatched = set(sam.index) - (set(tracker.index) | set(["nan"]))
+        if not addMissing:
```
Comment (jkobject) [inline depmapomics/terra.py:236]: same

File: depmapomics/terra.py [hunk 26]
```diff
@@ -226,5 +229,109 @@ def copyToWorkspace(
...
+            print("found these to be unmatched in the tracker: " + str(unmatched))
+            if deleteUnmatched and len(unmatched) > 0:
+                terra.removeSamples(workspaceID, unmatched)
+        unmatched = set(tracker.index) - set(sam.index)
+        if len(track) != 0:
+            for i in range(0, len(track), group):
+                wm.update_sample_attributes(track.iloc[i : i + group])
+        if addMissing and len(unmatched) > 0:
```
Comment (jkobject) [inline depmapomics/terra.py:244]: same

File: depmapomics/copynumbers.py (hunk 4 above)
Comment (5im1z) [reply depmapomics/copynumbers.py:473]: I don't remember this conversation but maybe I dropped it accidentally. I'll add it back

File: depmapomics/fingerprinting.py (hunk 10 above)
Comment (5im1z) [reply depmapomics/fingerprinting.py:8]: no not really haha

File: depmapomics/fingerprinting.py (hunk 15 above)
Comment (5im1z) [reply depmapomics/fingerprinting.py:57]: would it be necessary to add a _CCLEfingerprint wrapper and put all the taiga stuff under it?

File: depmapomics/config_dev.py [hunk 27]
```diff
@@ -304,6 +304,24 @@
 WESCNWORKSPACE = "broad-firecloud-ccle/DepMap_WES_CN_hg38"
 WESMUTWORKSPACE = "broad-firecloud-ccle/DepMap_Mutation_Calling_CGA_pipeline"
```
Comment (jkobject) [inline depmapomics/config_dev.py:305]: still did not change these.

File: depmapomics/config_dev.py [hunk 28]
```diff
@@ -304,6 +304,24 @@
...
+WORKING_DIR = "temp/"
+
+FPWORKSPACE = "broad-firecloud-ccle/CCLE_SNP_QC-copy"
+
+LEGACY_BAM_COLNAMES = ["legacy_bam_filepath", "legacy_bai_filepath"]
+
+FPALLBATCHPAIRSETS = "all"
+
+TAIGA_FP = "ccle-bam-fingerprints-6f30"
+TAIGA_FP_FILENAME = "fingerprint_lod_matrix.csv"
```
Comment (jkobject) [inline depmapomics/config_dev.py:322]: same here

File: depmapomics/fingerprinting.py [hunk 29]
```diff
@@ -0,0 +1,423 @@
...
+
+    Args:
+        wm (wmanager): workspace manager of current FP terra workspace
+        sampleset (str): name of current sample set
+        working_dir (str): directory where the lod matrix file will be saved
+        batch_pair_entity (str): name of entity on terra workspace that represents sample batch pairs
+        save_new_mat (bool): if new lod matrix (not combined with a previous matrix) is saved
+        new_mat_filename (str): name of the new lod matrix file if save_new_mat
+        prev_mat_df (pd.DatatFrame): a previous lod matrix file that will be merged with the new lod matrix
+        updated_mat_filename (str): name of the updated (new merged with prev) lod matrix file
```
Comment (jkobject) [inline depmapomics/fingerprinting.py:35]: need returns as well

File: depmapomics/fingerprinting.py [hunk 30]
```diff
@@ -0,0 +1,423 @@
...
+    updated_mat_filename="updated_lod_matrix.csv"
+):
+    """Here we update the fingerprint LOD matrix on taiga with the new fingerprints
+    and enerate matrix with LOD score for new fingerprint vcfs
+
+    Args:
+        wm (wmanager): workspace manager of current FP terra workspace
+        sampleset (str): name of current sample set
+        working_dir (str): directory where the lod matrix file will be saved
+        batch_pair_entity (str): name of entity on terra workspace that represents sample batch pairs
```
Comment (jkobject) [inline depmapomics/fingerprinting.py:31]:
```suggestion
        batch_pair_entity (str): optional. name of entity on terra workspace that represents sample batch pairs. Defaults to "sample_batch_pair".
```
when optional

File: depmapomics/fingerprinting.py [hunk 31]
```diff
@@ -0,0 +1,423 @@
...
+        save_new_mat (bool): if new lod matrix (not combined with a previous matrix) is saved
+        new_mat_filename (str): name of the new lod matrix file if save_new_mat
+        prev_mat_df (pd.DatatFrame): a previous lod matrix file that will be merged with the new lod matrix
+        updated_mat_filename (str): name of the updated (new merged with prev) lod matrix file
+    """
+
```
Comment (jkobject) [inline depmapomics/fingerprinting.py:37]:
```suggestion
```

File: depmapomics/fingerprinting.py [hunk 32]
```diff
@@ -0,0 +1,423 @@
...
+        )
+        updated_lod_mat = pd.concat(
+            (
+                updated_lod_mat.loc[new_ids.union(old_ids), old_ids],
+                new_lod_mat.transpose().loc[new_ids.union(old_ids, new_ids)],
+            ),
+            axis=1,
+        )
+        updated_lod_mat.to_csv(working_dir + updated_mat_filename + ".csv")
+
```
Comment (jkobject) [inline depmapomics/fingerprinting.py:77]:
```suggestion
```

File: depmapomics/fingerprinting.py [hunk 33]
```diff
@@ -0,0 +1,423 @@
...
+    return new_ids, updated_lod_mat
+
+
+def checkMismatches(lod_mat, ref, samples, thr=100):
```
Comment (jkobject) [inline depmapomics/fingerprinting.py:81]: still missing documentation everywhere else

File: depmapomics/fingerprinting.py [hunk 34]
```diff
@@ -0,0 +1,423 @@
...
+                                    "participant_id",
+                                    "stripped_cell_line_name",
+                                ],
+                            ]
+                            .astype(str)
+                            .values.tolist()
+                        ): n
+                    }
+                )
+            n = [
```
Comment (jkobject) [inline depmapomics/fingerprinting.py:178]: what is `n`? can we change its name?

File: depmapomics/fingerprinting.py [hunk 35]
```diff
@@ -0,0 +1,423 @@
...
+    v = updated_lod_mat.loc[new_ids]
+    ref = tracker.getTracker()
+    ref = ref.append(samples)
+
+    # find samples that should match but don't
+    mismatches = checkMismatches(v, ref, samples)
+
+    # find samples that shouldn't match but do
+    matches = checkMatches(v, ref)
+
```
Comment (jkobject) [inline depmapomics/fingerprinting.py:363]:
```suggestion
```

File: depmapomics/fingerprinting.py [hunk 36]
```diff
@@ -0,0 +1,423 @@
...
+    sampleset=SAMPLESETNAME,
+    allbatchpairset=FPALLBATCHPAIRSETS,
+    workspace=FPWORKSPACE,
+    working_dir=WORKING_DIR,
+    bamcolname=LEGACY_BAM_COLNAMES,
+    taiga_dataset=TAIGA_FP,
+    updated_mat_filename=TAIGA_FP_FILENAME
+):
+    """ CCLE fingerprinting function
+    Args:
```
Comment (jkobject) [inline depmapomics/fingerprinting.py:381]:
```suggestion
    
    Args:
```

File: depmapomics/fingerprinting.py [hunk 37]
```diff
@@ -0,0 +1,423 @@
...
+    await terra.waitForSubmission(workspace, submission_id)
+
+    # save config after jobs finish running
+    terra.saveWorkspace(workspace, "data/" + sampleset + "/FPconfig/")
+
+    # Update LOD matrix
+    new_ids, updated_lod_mat = updateLOD(wm, sampleset, working_dir, save_new_mat=False, prev_mat_df=None, updated_mat_filename=updated_mat_filename)
+
+    # finding issues with the dataset
+    v = updated_lod_mat.loc[new_ids]
```
Comment (jkobject) [inline depmapomics/fingerprinting.py:354]: change v too?

File: depmapomics/loading.py [hunk 38]
```diff
@@ -1174,7 +1174,7 @@ def update(
         add_to_samplesets (list, optional): add new samples to additional sample_sets on terra. Defaults to []
     """
     # uploading to our bucket (now a new function)
-    samples = terra.changeToBucket(
+    samples =  terra.changeToBucket(
```
Comment (jkobject) [inline depmapomics/loading.py:1177]:
```suggestion
    samples = terra.changeToBucket(
```

File: depmapomics/fingerprinting.py (hunk 37 above)
Comment (5im1z) [reply depmapomics/fingerprinting.py:352]: just to clarify, should the input for checkMismatches() and checkMatches() be just the portion of the lod matrix with the new lines? I think in 21Q4 v wasn't defined in the notebook so I assumed it was the whole matrix, but it makes more sense if it's just the slice with new ids

**Pull request merged**